src/main/scala/is/hail/expr/ir/TableIR.scala

```diff
@@ -453,6 +453,34 @@ case class TableJoin(left: TableIR, right: TableIR, joinType: String) extends Ta
   }
 }
 
+case class TableLeftJoinRightDistinct(left: TableIR, right: TableIR, root: String) extends TableIR {
+  require(left.typ.key.isDefined)
+  require(right.typ.key.isDefined)
+  require(left.typ.keyType.exists(l => right.typ.keyType.exists(r => l.isIsomorphicTo(r))))
```

Comment:
The last require implies the first two. Also, I think the precondition should be `r isPrefixOf l`.